File: solutions_python/parsons4.py
"""Now let's talk about climate change"""

import pandas as pd #dataframes
import numpy as np #number functions
import matplotlib.pyplot as plt #plotting functions
import os #file and folder functions

df=pd.read_csv(os.path.join('outputs','shefmetdat.csv')) #import cleaned data we exported from the previous script parsons3.py

print(df.dtypes) #check what data types the columns imported as 

max_temp_mean=df.groupby('mm')['temp_max'].mean() #average max temp for each month
max_temp_stdv=df.groupby('mm')['temp_max'].std() #stdev of max temp for each month

def zscorefunc(df,meanval,stdval): #function to calculate z scores
    return (df['temp_max']-meanval[df['mm']])/stdval[df['mm']] #return z score = val-mean/stv

args=(max_temp_mean,max_temp_stdv) #arguments we will pass to the function
df['ztempmax']=df.apply(zscorefunc,args=args,axis=1) #create new colum with z score of temp for each month

plt.clf() #clear any existing plots
df.groupby('mm')['ztempmax'].mean().plot() #sanity check - average z score should be close to zero

df2=df.groupby('yyyy')['ztempmax'].mean().reset_index()  # find average z score for each year
df2['5year_average']=df2['ztempmax'].rolling(center=False,window=5).mean() #create 5 year rolling average 

plt.clf() #clear any existing plots
plt.plot(df2['yyyy'],df2['ztempmax'],'.',color='b',ms=5,label='yearly average') #scatter plot of yearly averages
plt.plot(df2['yyyy'],df2['5year_average'],color='r',lw=3,label='5 year rolling mean') #line of rolling average
plt.ylabel('Average z score of max monthly temperature') #add y label
plt.xlabel('year') #add x label
plt.legend(loc=0) #add legend
plt.savefig(os.path.join('outputs','sheffield1883-2019.png')) #save figure
#conclusion = be terrified of climate change?

File: solutions_R/parsons4.R
"""MET office data for the sheffield station from https://www.metoffice.gov.uk/pub/data/weather/uk/climate/stationdata/sheffielddata.txt"""

import pandas as pd #dataframes
import numpy as np #number functions
import matplotlib.pyplot as plt #plotting functions
import os #file and folder functions

df=pd.read_csv(os.path.join('inputs','sheffieldmetdata_cleaned.txt'),skiprows=[1],sep='\s{1,}') #import data, using 1 or more space as the seperator

df.columns=['yyyy','mm','temp_max','temp_min','airfrost','rain','sun'] #rename the columns to be a bit more readable

df.replace('---',np.nan,inplace=True) #replace all the ---s with NaNs to indicate missing data

print(df.dtypes) #check what data types the columns imported as 

vars_to_convert=["temp_max", "temp_min","airfrost","sun"] #list of variables to change to numeric type

df[vars_to_convert] = df[vars_to_convert].apply(pd.to_numeric) #convert some variables to numeric type

print(df.dtypes) #check what data types the columns are now


plt.clf() #clean existing plot, if any
df.groupby('mm')['temp_max'].mean().plot(label='max',color='r') #plot average max temp for each month
df.groupby('mm')['temp_min'].mean().plot(label='min',color='b') #plot average max temp for each month
plt.legend(loc=0) #add legend
plt.xlabel('Month (1-12)') #xlabels
plt.ylabel('Average max/min tempreture in degree C') #ylabels
plt.savefig(os.path.join('outputs','minmaxtemp.png')) #export

df.to_csv(os.path.join('outputs','shefmetdat.csv')) #export prcoessed data to csv
